### ISSUE TITLE
Finish rewriting AminoREAPI

### ISSUE THREAD

**Issue by MrPowerGamerBR** [MrPowerGamerBR/AminoREAPI#2]
Currently the version on this repo is outdated.

I'm currently rewriting AminoREAPI for Vespertine (https://vespertine.mrpowergamerbr.com), my unofficial Amino client.

However it is nowhere near finished, but should be soon up on this repo.

**Comment by matatodo3000** [comment]
:relaxed:

**Comment by moelrobi** [comment]
If you need help, I would like to do it. :)

Your repo is the reason I created a node library for working with the Narvii-API.

**Comment by MrPowerGamerBR** [comment]
@moelrobi this project (Vespertine + AminoREAPI) is on hold because I got busy with other projects ( https://loritta.website/ ) (and I also got super mad how Amino doesn't use WebSockets for the chat stuff, even if the app ALREADY USES WebSockets for... guess what? typing status!)

My local version is already miles better than the version on this repo, I want to push it soon:tm: to here. (I also need to fix my Discord Bot's Amino post sync API)

What would be a good idea is to make a API documentation somewhere... hmmmm... 🤔

**Comment by MrPowerGamerBR** [comment]
owo whats this: https://github.com/MrPowerGamerBR/AminoREAPI/commit/c80c4b3f6dd46ee97230f21b7322aaa6a9a30eee

And here is how to interact with the WebSocket stuff (I didn't add anything that uses WebSockets yet, but I want to keep here how I did interact with the Amino WebSocket stuff)

``` 
val signBody = amino.deviceId + "|" + System.currentTimeMillis()
val url = "wss://ws2.narvii.com/?signBody=${URLEncoder.encode(signBody, "UTF-8")}"

val builder = Request.Builder().url(url).header("User-Agent", "Amino Client").header("NDCDEVICEID", amino.deviceId)
builder.header("NDCAUTH", "sid=${amino.sessionId}")

val socket = OkHttpClient().newWebSocket(builder.build(), AminoSocketListener())

val obj = JsonObject()
obj["ndcId"] = "aaa"
obj["topic"] = "shrug"
socket.send(gson.toJson(WsMessage(300, obj)))
```

**Comment by MrPowerGamerBR** [comment]
What got me a bit worried is that it seems Amino is going to start signing the packets too `NDC-MSG-SIG`, it seems not all packets do use the signature (only the packets that doesn't use the `NDC-AUTH` header) and for now I'm just generating a random UUID for the signature: https://github.com/MrPowerGamerBR/AminoREAPI/blob/master/src/main/java/com/mrpowergamerbr/aminoreapi/Amino.kt#L43

**Comment by MrPowerGamerBR** [comment]
By the way, if anyone wants to help, here is what I use to do Amino stuff:

1. IntelliJ IDEA
2. [Packet Capture](https://play.google.com/store/apps/details?id=app.greyshirts.sslcapture&hl=en) (if you can stand looking at your phone & fails to decode WebSocket traffic) and [Fiddler](https://www.telerik.com/fiddler)
3. [Frida](https://www.frida.re/) (I used it to dump what the app receives/sends with WebSockets, here's the code I used: https://gist.github.com/MrPowerGamerBR/69bd49ed92a517c78ff90493bf3abecc )
4. Patience
5. #BlameNarvii

**Comment by coffeeboo** [comment]
To be clear, does the "new" version actually include new endpoints that may have been added by Narvii, or is it just a better interface for Kotlin?

I'm trying to use all this for Swift and Node.JS anyway, so it doesn't matter 😄

**Comment by MrPowerGamerBR** [comment]
@coffeeboo not yet, I just rewrote (well, refactored it) to be better.

While I did implement new endpoints (and use some of the old endpoints) some of the old ones were removed and some of the old ones were changed to be better suited to Kotlin.

Some of nifty features were added too, like MediaType deserialization.

**Comment by moelrobi** [comment]
@MrPowerGamerBR well but is there any Clue what they are singing the Messages for? Or yet better how they do it? I mean I wouldn't see a reason why they would do that. :thinking: 

@coffeeboo i am already creating a Amino libary for interacting with Amino's API via Node. If you want to Helpout i can add you to the project

https://gitlab.robsys.space/Kikai/KikaiFramework

E-Mail Me: moelrobi@gmail.com if you are interested.

**Comment by MrPowerGamerBR** [comment]
@moelrobi no, I'm not sure why they are signing the requests, wouldn't it be easier if they just switched to HTTPS? 🤔

**Comment by MrPowerGamerBR** [comment]
I think it would be a good idea to start documenting the Amino API to keep everything nice and tidy, examples:

https://www.schiff.io/projects/humble-bundle-api
http://wiki.vg/Protocol
https://soupply.github.io/

**Comment by coffeeboo** [comment]
@MrPowerGamerBR HOLD UP. Narvii doesn't use HTTPS?

**Comment by moelrobi** [comment]
@coffeeboo thier Endpoints resolve after http://service.narvii.com/api/v1

they are using ws and HTTPv.1.0

**Comment by MrPowerGamerBR** [comment]
@coffeeboo no they don't, but they do support HTTPS, which is... strange, very strange.

**Comment by mclainj** [comment]
I apologize for the unrelated comment, but I was unsure of where to ask this (not much experience in GitHub).
I'm looking to do some natural language processing and analysis on blogs posted within an Amino community. Is there currently a way to access this kind of data? I do not have much experience dealing with APIs, so I'm sorry for the ignorance.

**Comment by MrPowerGamerBR** [comment]
Yes it is possible to pull posts from a community

Em sex, 6 de jul de 2018 23:13, McLain Johnson <notifications@github.com>
escreveu:

**Comment by mclainj** [comment]
Thanks for the quick response! Is that only possible through AminoREAPI at the moment? I read that an Android device was required, and I'm afraid I lack that.

**Comment by moelrobi** [comment]
Well Node got Two Libarys, Java / Kotlin One and there is one C# one.

**Comment by MrPowerGamerBR** [comment]
Now this project is on hold, the reason I was reverse engineering the API was to write a unofficial web client for Amino... it was like, my motivation to reverse engineer it...

...but now Amino has an official web client 😭

**Comment by MrPowerGamerBR** [comment]
*Maybe* I will continue the project if, one day, I port my Discord bot to Amino... maybe... who knows!

**Comment by MrPowerGamerBR** [comment]
If you have Android Nougat+, you will need this Magisk module to intercept the connection with Fiddler/Packet Capturer: https://github.com/NVISO-BE/MagiskTrustUserCerts

Requires Magisk, I tested with Amino and it works fine :)

**Comment by MrPowerGamerBR** [comment]
![https://cdn.discordapp.com/attachments/358774895850815488/533083850339909632/Screenshot_20190110-224443.png](https://cdn.discordapp.com/attachments/358774895850815488/533083850339909632/Screenshot_20190110-224443.png)

Playing around while I'm recoding the API :3

**Comment by MrPowerGamerBR** [comment]
Looks like Amino now uses WebSockets for message receiving (yay!)

Reverse engineering that is way easier, since Amino Web also uses WebSockets and looking at the WebSocket messages is easy using Chrome (in the network tab -> "WS" -> select the connection -> Messages)

**Comment by MrPowerGamerBR** [comment]
I decided to push all my old changes to the `master` branch, it is not ready, but it does work: https://github.com/MrPowerGamerBR/AminoREAPI/commit/866c4f7d01c2f7a32006199cd2d30044665ad86a

**Comment by MrPowerGamerBR** [comment]
You can now listen for incoming messages via WebSockets.

```kotlin
 val client = AminoClientBuilder()
                .setDeviceId("DeviceID")
                .withCredientials("Email", "Password")
                .connectToWebSocket(true)
                .enableCache(true)
                .addEventListener(MessageListener())
                .connect()
```

```kotlin
class MessageListener : ListenerAdapter() {
    override suspend fun onMessageReceived(event: MessageReceivedEvent) {
        println("Message received! ${event.message.author.nickname}: ${event.message.content}")
    }
}
```

**Comment by MrCreep** [comment]
I still do not understand how to create a bot.  Can someone help describe in stages how to do?  Sorry for being so stupid: d
